optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/AbstractScoreDirector.java

```diff
@@ -172,9 +172,20 @@ public SupplyManager getSupplyManager() {
     // Complex methods
     // ************************************************************************
 
+    private void assertWorkingSolutionClassSameAsAnnotatedClass(Solution_ workingSolution) {
+        Class<Solution_> actualSolutionClass = (Class<Solution_>) Objects.requireNonNull(workingSolution).getClass();
+        Class<Solution_> expectedSolutionClass = getSolutionDescriptor().getSolutionClass();
+        if (!actualSolutionClass.equals(expectedSolutionClass)) {
```

Comment:
The workingSolution is also allowed to be a subclass of the annotated class (but not the other way around).
Use `Class.isAssignableFrom`.

**Why doesn't TestdataUnannotatedExtendedSolution trigger this fail fast?**

Reply:
It does, which is why the test was removed.